Terraform crash on empty module source

### Terraform Version

```shell
latest
```

### Terraform Configuration Files

```terraform
terraform {
  required_version = ">= 1.14"

  required_providers {
    local = {
      source  = "hashicorp/local"
      version = "~> 2.7"
    }
  }
}

module "bug" {
  version = "1"
}
```


### Debug Output

```
terraform init
Initializing modules...

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: runtime error: invalid memory address or nil pointer dereference
goroutine 25 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:26 +0x64
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x16c
panic({0x107bc8640?, 0x10a8a5820?})
        runtime/panic.go:783 +0x120
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1.1()
        github.com/hashicorp/terraform/internal/terraform/graph.go:59 +0x30c
panic({0x107bc8640?, 0x10a8a5820?})
        runtime/panic.go:783 +0x120
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0x14000736000, {0x0, 0x0}, {{0x10847c2d8?, 0x1076c19b0?}})
        github.com/hashicorp/terraform/internal/lang/eval.go:180 +0x148
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateExpr(0x0?, {0x0, 0x0}, {{0x10847c2d8?, 0x1076c19b0?}}, {0x0?, 0x0?})
        github.com/hashicorp/terraform/internal/terraform/eval_context_builtin.go:353 +0x84
github.com/hashicorp/terraform/internal/terraform.evalSource({0x0, 0x0}, 0x1, {0x1084b1fc0, 0x14000829680})
        github.com/hashicorp/terraform/internal/terraform/node_module_install.go:188 +0x350
github.com/hashicorp/terraform/internal/terraform.(*nodeInstallModule).Execute(0x14000471ef0, {0x1084b1fc0, 0x14000829680}, 0xc0?)
        github.com/hashicorp/terraform/internal/terraform/node_module_install.go:87 +0x1b0
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0x1400085a000, {0x1084b1fc0, 0x14000829680}, {0x111ca42c0, 0x14000471ef0})
        github.com/hashicorp/terraform/internal/terraform/graph_walk_context.go:170 +0xa0
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x107f1a0a0, 0x14000471ef0})
        github.com/hashicorp/terraform/internal/terraform/graph.go:143 +0x598
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0x1400083dda0, {0x107f1a0a0, 0x14000471ef0}, 0x14000819300)
        github.com/hashicorp/terraform/internal/dag/walk.go:393 +0x284
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 1
        github.com/hashicorp/terraform/internal/dag/walk.go:316 +0xae4
```


### Expected Behavior

No panic

### Actual Behavior

paaaanic

### Steps to Reproduce

`terraform init`

### Additional Context

Related:
* https://github.com/hashicorp/terraform/pull/38560
* https://github.com/hashicorp/terraform/pull/38227/changes#diff-c53dd53740c81e9b2640a4013b03ba28088c2d1b90ca9c9014be7b368cd4f6efR181-R183

### References

_No response_

### Generative AI / LLM assisted development?

_No response_

--- a/.changes/v1.15/BUG FIXES-20260520-183443.yaml
+++ b/.changes/v1.15/BUG FIXES-20260520-183443.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: Fix crash on `init` for modules with empty source
+time: 2026-05-20T18:34:43.331672+02:00
+custom:
+    Issue: "38628"

--- a/internal/command/init_test.go
+++ b/internal/command/init_test.go
@@ -6374,6 +6374,36 @@ func TestInit_varValueWithoutConfig(t *testing.T) {
 	}
 }
 
+func TestInit_invalidConfig(t *testing.T) {
+	td := t.TempDir()
+	t.Chdir(td)
+	cfg1 := `module "test" { version = "1" }`
+
+	if err := os.WriteFile(filepath.Join(td, "main.tf"), []byte(cfg1), 0644); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	ui := cli.NewMockUi()
+	view, done := testView(t)
+	initCmd := &InitCommand{
+		Meta: Meta{
+			Ui:   ui,
+			View: view,
+		},
+	}
+
+	code := initCmd.Run([]string{})
+	testOutput := done(t)
+	if code != 1 {
+		t.Fatalf("expected init to fail with code 1, got code %d: \n%s", code, testOutput.All())
+	}
+
+	expectedErr := "Error: Missing required argument"
+	if !strings.Contains(cleanString(testOutput.Stderr()), expectedErr) {
+		t.Fatalf("unexpected error, expected %q, given: %s", expectedErr, testOutput.Stderr())
+	}
+}
+
 // newMockProviderSource is a helper to succinctly construct a mock provider
 // source that contains a set of packages matching the given provider versions
 // that are available for installation (from temporary local files).

--- a/internal/terraform/node_module_install.go
+++ b/internal/terraform/node_module_install.go
@@ -73,6 +73,12 @@ func (n *nodeInstallModule) References() []*addrs.Reference {
 func (n *nodeInstallModule) Execute(ctx EvalContext, walkOp walkOperation) tfdiags.Diagnostics {
 	var diags tfdiags.Diagnostics
 
+	if n.ModuleCall.SourceExpr == nil {
+		// Cannot install a module without a source
+		// Outer diags should already contain a parsing error
+		return diags
+	}
+
 	var version configs.VersionConstraint
 	if n.ModuleCall.VersionExpr != nil {
 		var versionDiags tfdiags.Diagnostics
@@ -139,6 +145,12 @@ func (n *nodeInstallModule) Execute(ctx EvalContext, walkOp walkOperation) tfdia
 func (n *nodeInstallModule) DynamicExpand(ctx EvalContext) (*Graph, tfdiags.Diagnostics) {
 	var g Graph
 	var diags tfdiags.Diagnostics
+
+	if n.Config == nil {
+		// Cannot expand without a config. This can happen when something goes wrong
+		// during module installation/Execute() above.
+		return nil, diags
+	}
 
 	expander := ctx.InstanceExpander()
 	_, call := n.Addr.Call()

